### ISSUE TITLE
Create a web server

### ISSUE THREAD

**Issue by tfukaza** [tfukaza/harvest#19]
One idea I had was to build a web server into Harvest. Users can access the server via a web browser to get the latest stats of their algorithm, like current positions and returns on investment. This will be a long-term project, and there are lots of specs to decide before starting:
- What framework to use for the web server
- What framework to use for the web interface returned by the server
- Possibility of elaborate GUI interfaces

**Comment by shershah010** [comment]
Crazy enough, I had the same idea. It could be like splunk or jupyter notebooks even. 

As for frameworks I would say flask or django since it is in python and we can easily import harvest modules. I am hesitant on having a frontend + backend setup because that is two servers instead of one.

**Comment by tfukaza** [comment]
My thought was that a frontend+backend setup would be better since the web interface would be like a web app. Also we can serve the website and the backend API from the same server, just from different endpoints (I think Vercel does something like this).

**Comment by shershah010** [comment]
In that case a react + flask / django flow seems good to me but I'll let you decide what tools to use.

**Comment by tfukaza** [comment]
I think Flask + React will be good :)

**Comment by tfukaza** [comment]
> Hi Guys, very impressive code design, same design I`ve started 6 months back and stopped. anyway my suggestion how about doing a mobile app with Qt/qml or google flutter, with server api as backend.

Yeah some kind of remote access from devices other than desktops/laptops will definitely be interesting to implement in the future. As for the tech stack, I think just using web frameworks will be easier since building a native app will be a bit beyond the scope of this project.

**Comment by KeoniGarner** [comment]
Hello there, I’m interested in this project. I noticed that Flask seems to be the prevailing framework. Don’t want to derail this too much, but I’m wondering if Django might be a better fit since there has been little discussion (from what I can tell) around DB management. Django’s ORM is quite nice, but I’m a bit biased because that’s what I work with full time.

If it’s decided that Flask will be used, what DB will we use and how will that be provisioned?

**Comment by tfukaza** [comment]
Thanks for your interest :) One motivation for using Flask is because Harvest doesn't really have a complex database system, so a simple web framework will suffice. Unlike a typical website that may have to serve millions of requests and store terabytes of user data, Harvest is mainly a client-side software, meaning its web server will only serve one user, and store a single user's data at a time. So a framework like Django will likely be an overkill. 

As for the database, the data will mainly be served from the `Storage` class. To give a quick overview, the `Storage` class is a wrapper around various data storage mediums (RAM, file system, SQL DB, etc.) which gives a common interface no matter where the user decides to store data. So in some sense, this class will be our "database".

**Comment by KeoniGarner** [comment]
Interesting… I like the `Storage` class concept. I’m wondering, then, if it might be cleaner **not** to create a web server for the client, but make it easily extensible - similar to how the `Storage` class works. That way the client can easily spin up web servers at their own discretion. Thoughts?

**Comment by tfukaza** [comment]
Thats true, but Harvest also targets people who are beginners in coding, so we can't assume the end users will know how to make their own web server. So for now, we'll need to make a web server/web interface. 

You bought up an interesting point though, it might be interesting to formally define a public API for the Storage class so people who _are_ experienced in coding can make their own software that interfaces with Harvest.

**Comment by KeoniGarner** [comment]
Okay fair enough - so since Harvest is client-focused, we will not actually provision a web server right? If I'm understanding correctly, it sounds like it will just be a local server for logs/monitoring.

FWIW (which probably isn't much since I'm new here) I like the idea of utilizing the `Storage` class in this manner

**Comment by tfukaza** [comment]
Yeah you got the idea 👍, this "web server" is just a fancy way to display local system log and trading stats.
